text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+There is an Authorizations Pallet which stores the association. Adding a new piece of code is permissionless but requires a deposit commensurate with its size.
+
+```rust
+type AuthId = u32;
+type Authorizers = StorageMap<AuthId, Authorizer>;
+```
+
+An *Authorization* is simply a blob which helps the Authorizer recognize a properly authorized Work Package. No constraints are placed on Authorizers over how they may interpret this blob. Expected authorization content includes signatures, Merkle-proofs and more exotic succinct zero-knowledge proofs.
+
+_(Note: depending on future Relay-chain Coretime scheduling implementation concerns, a window of Relay-chain blocks)._
+
+The need of validators to be rewarded for doing work they might reasonably expect to be useful competes with that of the Coretime procurers to be certain to get work done which is useful to them. In Polkadot 1.0, validators only get rewarded for PoVs ("work packages") which do not panic or overrun. This ensures that validators are well-incentivized to ensure that their computation is useful for the assigned parachain. This incentive model works adequately where all PVF code is of high quality and collators are few and static.
+
+However with this proposal (and even the advent of on-demand parachains), validators have little ability to identify a high-quality Work Package builder and the permissionless design means a greater expectation of flawed code executing in-core. Because of this, we make a slightly modified approach: Work Packages must have a valid Authorization, i.e. the Coretime-assigned `is_authorized` returns `true` when provided with the Work Package. However, Validators get rewarded for *any* such authorized Work Package, even one which ultimately panics or overruns on its evaluation.
+
+This ensures that Validators do a strictly limited amount of work before knowing whether they will be rewarded and are able to discontinue and attempt other candidates earlier than would otherwise be the case. There is the possibility of wasting Coretime by processing Work Packages which result in error, but well-written authorization procedures can mitigate this risk by making a prior validation of the Work Items.
+
+### Refine
+
+The `refine` function is implemented as an entry-point inside a code blob which is stored on-chain and whose hash is associated with the Work Class.
+
+```rust
+type ClassCodeHash = StorageMap<ClassId, CodeHash>;
+```
+
+```rust
+type WorkOutputLen = ConstU32<4_096>;
+type WorkOutput = BoundedVec<u8, WorkOutputLen>;
+fn refine(
+    payload: WorkPayload,
+    authorization: Authorization,
+    auth_id: Option<AuthId>,
+    context: Context,
+    package_hash: WorkPackageHash,
+) -> WorkOutput;
+```
```

Comment:
Now I have the opposite comment to the one I had earlier. So refine doesn't depend on the state of anything, unless it was proved from the headerhash, which is definitely inefficient. That means that state dependencies e.g. parachain head data and relay chain state root dependencies must be given as hashes in the WorkOutput, which are then checked by accumulate? But then, for relay chain data, we would need to check past data, because this data may have changed. Maybe refine should have more relay chain things as input through the Context or otherwise.

Reply:
I added state root and beefy root as per discussion.

Reply:
> parachain head data and relay chain state root dependencies must be given as hashes in the WorkOutput

We'd need to do this anyways to do multiple parachian blocks per relay chain block, no?